Support of multiple persistence units for Hibernate Panache

machi1990 opened the pull request:
As proposed by Guillaume in this comment https://github.com/quarkusio/quarkus/issues/2759#issuecomment-531191748

Supersedes https://github.com/quarkusio/quarkus/pull/4018 (which was quite old)

/cc @gsmet @Sanne

machi1990 committed:
feat(panache): add support of multiple persistence units in hibernate
orm

gsmet commented:
This is on ice for now, it's part of the larger multiple persistence units work.

Sanne commented:
@gsmet please elaborate? I should not review this?

machi1990 commented:
@Sanne we'll need multiple persistence unit support for the hibernate extension. So let's keep this on-ice for now as Guillaume suggested.

gastaldi commented:
@gsmet I believe you're working on this now? Maybe we can close this PR?

gsmet commented:
Please let it open for now. I’m sure Manyanda will get back to it once I’m done with the Hibernate ORM extension part. 

I’m keeping track of things, don’t worry.

machi1990 commented:
@gsmet just rebased this PR on top of https://github.com/quarkusio/quarkus/pull/11648. Let's get that one in first and then this afterwards.

gsmet commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:
```diff
@@ -10,20 +10,8 @@
-import java.util.List;
-import java.util.Locale;
-import java.util.Map;
+import java.util.*;
```
No star imports, please.

machi1990 replied:
Good catch. Huh this should have been gone.

gsmet commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:
```diff
@@ -255,7 +243,8 @@ public void configurationDescriptorBuilding(
             BuildProducer<HotDeploymentWatchedFileBuildItem> hotDeploymentWatchedFiles,
-            BuildProducer<PersistenceUnitDescriptorBuildItem> persistenceUnitDescriptors) {
+            BuildProducer<PersistenceUnitDescriptorBuildItem> persistenceUnitDescriptors,
+            BuildProducer<JpaEntityPerPersistenceUnitsBuildItem> jpaEntitiesPerPersistenceUnits) {
```
I think I would name it `JpaModelPersistenceUnitMappingBuildItem`.

Also I already built something here to push it to the JPAConfig: https://github.com/quarkusio/quarkus/blob/master/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L455 so I don't think we should build it again .

It's not perfect either as we can't really properly deal with XML descriptors but I think it's good enough for what we want.

machi1990 replied:
> I think I would name it `JpaModelPersistenceUnitMappingBuildItem`.
> 
> Also I already built something here to push it to the JPAConfig: https://github.com/quarkusio/quarkus/blob/master/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L455 so I don't think we should build it again .
> 

Ah that's better, for some reason I missed this. I'd re-use this. 

> It's not perfect either as we can't really properly deal with XML descriptors but I think it's good enough for what we want.

In case of XML descriptors, what is the best way to deal with them? Should we use the default PU or ?

gsmet replied:
Well, have a look if it works or not but I think you won't have the entities in the descriptors.

Frankly, I think we don't really care, at least not in this first version of the multi-persistence unit support.

machi1990 replied:
Okay.

gsmet commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:
```diff
@@ -269,10 +258,19 @@ public void configurationDescriptorBuilding(
+            for (String entityClassName : jpaEntities.getAllModelClassNames()) {
+                entityToPersistenceUnit.put(entityClassName, Arrays.asList(PersistenceUnitUtil.DEFAULT_PERSISTENCE_UNIT_NAME)); // use the default entity manager
+            }
+            jpaEntitiesPerPersistenceUnits.produce(new JpaEntityPerPersistenceUnitsBuildItem(entityToPersistenceUnit));
```
Yeah, so in this case, we basically can't really say the mapping, so maybe I would just avoid producing the build item.

Especially since it's not obvious you will produce only one of this (you can have XML descriptors and a config persistence unit).

gsmet commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaEntityPerPersistenceUnitsBuildItem.java:
```diff
@@ -0,0 +1,22 @@
+import io.quarkus.builder.item.SimpleBuildItem;
+
+/**
+ * Internal model to hold a Jpa entity to its corresponding persistence units.
```
```suggestion
 * Internal model to hold the mapping linking a JPA entity to its corresponding persistence units.
```

gsmet commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaEntityPerPersistenceUnitsBuildItem.java:
```diff
@@ -0,0 +1,22 @@
+    private final Map<String, List<String>> entityToPersistenceUnits;
+
+    public JpaEntityPerPersistenceUnitsBuildItem(Map<String, List<String>> entityToPersistenceUnits) {
+        this.entityToPersistenceUnits = entityToPersistenceUnits;
```
Maybe make it an unmodifiableMap?

machi1990 replied:
Sure.

gsmet commented on extensions/panache/hibernate-orm-panache-kotlin/deployment/src/main/java/io/quarkus/hibernate/orm/panache/kotlin/deployment/KotlinPanacheResourceProcessor.java:
```diff
@@ -185,5 +201,24 @@ void build(CombinedIndexBuildItem index,
+            boolean isPanacheEntity = modelClasses.stream().anyMatch(name -> name.equals(entityName));
+            if (selectedPersistenceUnits.size() > 1 && isPanacheEntity) {
+                throw new IllegalStateException(String.format(
+                        "PanacheEntity '%s' cannot be defined for usage in 'several persistence unit. The following persistence units were found %s'",
```
```suggestion
                        "PanacheEntity '%s' cannot be defined for usage in several persistence units. The following persistence units were found: %s.",
```

gsmet commented on extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheHibernateResourceProcessor.java:
```diff
@@ -194,4 +202,31 @@ void build(CombinedIndexBuildItem index,
+
+                if (selectedPersistenceUnits.size() > 1 && isPanacheEntity) {
+                    throw new IllegalStateException(String.format(
+                            "PanacheEntity '%s' cannot be defined for usage in 'several persistence unit. The following persistence units were found %s'",
```
```suggestion
                            "PanacheEntity '%s' cannot be defined for usage in several persistence units. The following persistence units were found: %s.",
```

gsmet commented on extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheHibernateResourceProcessor.java:
```diff
@@ -194,4 +202,31 @@ void build(CombinedIndexBuildItem index,
+
+                if (selectedPersistenceUnits.size() > 1 && isPanacheEntity) {
+                    throw new IllegalStateException(String.format(
+                            "PanacheEntity '%s' cannot be defined for usage in 'several persistence unit. The following persistence units were found %s'",
```
Also it's not very clear in the message that this is not supported.

machi1990 replied:
Added a few words about this.

gsmet commented on extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/JpaOperations.java:
```diff
@@ -71,16 +85,31 @@ public static boolean isPersistent(Object entity) {
-        delegate.flush();
+        Set<String> values = new HashSet<>(entityToPersistenceUnit.values());
+        for (String persistentUnit : values) {
+            getEntityManager(persistentUnit).flush();
```
Shouldn't we only flush the entityManager of the entity?

machi1990 replied:
Yes, this is already done for https://github.com/quarkusio/quarkus/pull/8695/files#diff-2fcdac996e03f2fce93a3bc400488d93R70

The method is poorly named. I think it should be `flushAll`, there is a TODO comment in the code. 

https://github.com/quarkusio/quarkus/pull/8695/files#diff-2fcdac996e03f2fce93a3bc400488d93R103

machi1990 replied:
Or maybe I am missing something.

gsmet replied:
Hmmm, I think the comment is on the wrong method then.

I think I would make `flush()` only flush the default persistence unit similar to what you did with other methods that are targeting the default if used without a parameter.

machi1990 replied:
So there are two methods:
- JpaOperations.flush(entity): this only flush the EntityManager of the entity
- JpaOperations.flush(): This flushes all the EntityManager. I was thinking that this one could be renamed to something like `flushAll()`, and certainly did not want to embark myself on that "refactoring things" journey.

gsmet replied:
I would keep the current behavior and flush the default entity manager.

We can add a flushAll later if someone asks for it.

machi1990 replied:
Sure.

machi1990 replied:
@gsmet this is now done.

gsmet commented on extensions/panache/hibernate-orm-panache-kotlin/runtime/src/main/java/io/quarkus/hibernate/orm/panache/kotlin/runtime/JpaOperations.java:
```diff
@@ -307,6 +330,13 @@ public static int update(Class<?> entityClass, String query, Object... params) {
+        public EntityManager getEntityManager(Class<?> clazz) {
+            String clazzName = clazz.getName();
+            String persistentUnitName = entityToPersistenceUnit.get(clazzName);
+            return super.getEntityManager(persistentUnitName);
```
I wonder if we could avoid a lot of overhead by having a generated method in the PanacheEntity/Repository return a static Literal for consumption by CDI.

It would avoid the overhead of the map lookup and would also avoid the overhead of instantiating a new Literal each time we want to the get EntityManager.

I'm not that versed into Panache so it might just be a crazy idea.

machi1990 replied:
Ah that's something I never thought of. This way seemed easier for me. I can look at that idea.

gsmet replied:
And for the default one, we would use `javax.enterprise.inject.Default.Literal.INSTANCE`.

gsmet replied:
Maybe let's get this first version in and then we work on a second iteration with code generation?

gsmet replied:
Depends on how fast you will be but I think Panache generation uses ASM so it won't be fun :).

machi1990 replied:
Sure, I'll focus on resolving other comments (https://github.com/quarkusio/quarkus/pull/8695#discussion_r478451025) first.

gsmet reviewed (changes requested):
Thanks for this work. I added some comments, all open to discussion, especially given I don't know Panache that much.

gsmet commented:
@machi1990 I merged my PR and this one should be rebased. Can you ping me when you're done with the comments I made (except for the code generation one) so that I have another look?

Thanks!

machi1990 commented:
> @machi1990 I merged my PR and this one should be rebased. Can you ping me when you're done with the comments I made (except for the code generation one) so that I have another look?
> 
> Thanks!

Hi @gsmet I was in the middle of the process and this is now done. I've pushed an update and also rebased. Let me know what you think.

gsmet committed:
Make JpaModelPersistenceUnitMappingBuildItem map unmodifiable

gsmet committed:
Improve error message for Panache entities attached to several PUs

gsmet committed:
Don't make the PU map unmodifiable as we clear it on destroy

gsmet committed:
Add the whole hierarchy of an entity to the persistence unit

gsmet commented on extensions/panache/hibernate-orm-panache-kotlin/deployment/src/main/java/io/quarkus/hibernate/orm/panache/kotlin/deployment/KotlinPanacheResourceProcessor.java:
```diff
@@ -185,5 +202,24 @@ void build(CombinedIndexBuildItem index,
+
+            panacheEntityToPersistenceUnit.put(entityName, selectedPersistenceUnits.get(0));
+        }
+        recorder.setEntityToPersistenceUnit(panacheEntityToPersistenceUnit);
```
Is this really needed? I would have expected the standard ORM processor to be triggered too? Or am I missing something?

gsmet replied:
If it's necessary, I will need to make the error message consistent but I would be surprised it is.

machi1990 replied:
> Is this really needed? I would have expected the standard ORM processor to be triggered too? Or am I missing something?

Yes, this is needed. The kotlin panache extension does not make use of the hibernate panache extension. They all share a common spi module.

gsmet replied:
Ok, I will adjust the PR then. Running the tests, I will push once they're green.

gsmet commented:
@machi1990 could you have a look to the commits I added? Also, I added a comment regarding the Kotlin part. Can you check?

Thanks!

gsmet commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:
```diff
@@ -956,6 +968,27 @@ private void enhanceEntities(final JpaEntitiesBuildItem domainObjects,
+        while (modelClassInfo != null && !modelClassInfo.name().equals(DotNames.OBJECT)) {
+            if (knownModelClassNames.contains(modelClassName)) {
+                relatedModelClassNames.add(modelClassInfo.name().toString());
+            }
```
It would probably be better to move this code just below the line below, otherwise we consider the current class too. It's not really a problem but it's less correct.

@machi1990 I'll wait to see if you have other comments before fixing this.

gsmet replied:
Just pushed the fix to have CI running this evening.

For me it looks ready to go, waiting for your blessing.

machi1990 replied:
I was just reviewing this part. The initial version would add the model to the `relatedModelClassNames` set too. Thanks for taking care of that.

machi1990 commented:
> @machi1990 could you have a look to the commits I added? Also, I added a comment regarding the Kotlin part. Can you check?
> 
> Thanks!

Reviewed recent changes to this and they look goof to me. The [only thing](https://github.com/quarkusio/quarkus/pull/8695#discussion_r480243051) that I was going to comment about has been fixed.

gsmet commented:
Nice work @machi1990 ! Let's try to follow up on the code generation before the Final so that the solution is fully optimized.

machi1990 commented:
I've that in my TODO list.
